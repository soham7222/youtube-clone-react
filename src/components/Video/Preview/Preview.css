.video_preview {
    display: grid;
    grid: 250px auto / 350px;
    margin: 10px;
}

.video_image {
    position: relative;
    grid-row: 1/2;
    grid-column: 1/2;
    cursor: pointer;
}

.video_preview_image {
    height: 250px;
    width: 350px;
    border-radius: 10px;
}

.video_preview_image:hover {
    border-radius: 0px;
    transition: 1s;
}

.video_timestamp {
    position: absolute;
    bottom: 0;
    right: 0;
    opacity: 0.8;
    border-radius: 2px;
    font-weight: 500;
    margin: 4px;
    padding: 2px 2px;
    line-height: 10px;
    background: black;
    color: white;
}

.video_info {
    grid-row: 2/3;
    grid-column: 1/2;
}

.video_channel_logo {
    width: 45px;
    height: 45px;
    border: 2px solid gold;
    background-color: beige;
    border-radius: 50%;
    margin: 5px;
    float: left;
    display: flex;
    justify-content: center;
    align-items: center;
    color: brown;
}

.verticalList {
    padding-left: 5px;
    grid-column: 2/3;
    grid-row: 1/2;
}

.video_basic_info {
    padding-top: 5px;
}

.video_channel {
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
    color: rgb(111, 100, 85)
}

.video_view_time {
    margin-top: 0px;
    margin-left: 59px;
    color: rgb(111, 100, 85)
}

.video_info_title {
    display: -webkit-box;
    overflow: hidden;
    text-overflow: ellipsis;
    -webkit-line-clamp: 2;
    -webkit-box-orient: vertical;
    line-height: 1.2em;
    max-height: 2.4em;
    font-family: 'Gill Sans', 'Gill Sans MT', Calibri, 'Trebuchet MS', sans-serif;
    margin-top: 5px;
    font-size: 18px;
}