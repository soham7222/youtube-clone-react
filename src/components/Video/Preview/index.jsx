export {
    Preview as PreviewComponent
} from "./Preview"