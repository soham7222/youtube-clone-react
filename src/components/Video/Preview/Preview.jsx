import React from 'react'
import "./Preview.css"

export const Preview = (props) => {
    const titleEl = document.createElement('div');
    titleEl.innerHTML = props.title;
    const plainTitle = titleEl.innerText;

    const getDuration = () => {
        const targetDate = new Date(props.publishTime);
        const currentDate = new Date(Date.now());
        const timeDiff = Math.abs(targetDate.getTime() - currentDate.getTime());
        const closestDuration = Math.round(timeDiff / (1000 * 60 * 60 * 24));
        if (closestDuration > 365) {
            const year = Math.round(closestDuration / 365)
            if (Math.round(closestDuration / 365) > 1) {
                return year + " years ago"
            } else {
                return year + " year ago"
            }
        }

        if (closestDuration > 30) {
            const month = Math.round(closestDuration / 30)
            if (Math.round(closestDuration / 365) > 1) {
                return month + " months ago"
            } else {
                return month + " month ago"
            }
        }

        if (closestDuration == 0) {
            const minutes = Math.round(timeDiff / (1000 * 60 * 60))
            return minutes + " minutes ago"
        }

        return closestDuration + " days ago"
    }



    return (
        <div className="video_preview">
            <div className="video_image">
                <img className="video_preview_image" src={props.image} />
                <div className="video_timestamp">
                    <span>{Math.floor(Math.random() * 60) + 1}:{Math.floor(Math.random() * 60) + 10}</span>
                </div>
            </div>
            <div className="video_info">
                <div className="video_channel_logo">
                    {props.channelName.slice(0, 2).toUpperCase()}
                </div>
                <div>
                    <div title={props.title} className="video_info_title">{plainTitle}</div>
                    <div className="video_basic_info">
                        <div className="video_channel">{props.channelName}</div>
                        <div className="video_view_time">{Math.floor(Math.random() * 20) + 1}k views ‚óè {getDuration()}</div>
                    </div>
                </div>
            </div>
        </div>
    )
}
