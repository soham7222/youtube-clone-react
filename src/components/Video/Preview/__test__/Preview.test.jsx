import React from 'react';
import { render } from '@testing-library/react';
import { Preview } from '../Preview';

describe('Preview Component', () => {
    beforeEach(() => {
        jest.spyOn(global.Date, 'now').mockImplementationOnce(() =>
            new Date('2019-05-14T11:01:58.135Z').valueOf()
        );
        jest.spyOn(global.Math, 'random').mockReturnValue(0.123456789);
    });

    afterEach(() => {
        jest.spyOn(global.Math, 'random').mockRestore();
    })

    it('renders correctly', () => {
        const dom = render(<Preview
            key={1}
            image="https://blog.hubspot.com/hubfs/image8-2.jpg"
            title="Google"
            channelName="Google Channel" />)
        expect(dom).toMatchSnapshot();
    });
})