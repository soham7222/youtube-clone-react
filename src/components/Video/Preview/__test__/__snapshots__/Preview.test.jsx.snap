// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Preview Component renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="video_preview"
      >
        <div
          class="video_image"
        >
          <img
            class="video_preview_image"
            src="https://blog.hubspot.com/hubfs/image8-2.jpg"
          />
          <div
            class="video_timestamp"
          >
            <span>
              8
              :
              17
            </span>
          </div>
        </div>
        <div
          class="video_info"
        >
          <div
            class="video_channel_logo"
          >
            GO
          </div>
          <div>
            <div
              class="video_info_title"
              title="Google"
            />
            <div
              class="video_basic_info"
            >
              <div
                class="video_channel"
              >
                Google Channel
              </div>
              <div
                class="video_view_time"
              >
                3
                k views ● 
                NaN days ago
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="video_preview"
    >
      <div
        class="video_image"
      >
        <img
          class="video_preview_image"
          src="https://blog.hubspot.com/hubfs/image8-2.jpg"
        />
        <div
          class="video_timestamp"
        >
          <span>
            8
            :
            17
          </span>
        </div>
      </div>
      <div
        class="video_info"
      >
        <div
          class="video_channel_logo"
        >
          GO
        </div>
        <div>
          <div
            class="video_info_title"
            title="Google"
          />
          <div
            class="video_basic_info"
          >
            <div
              class="video_channel"
            >
              Google Channel
            </div>
            <div
              class="video_view_time"
            >
              3
              k views ● 
              NaN days ago
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
